File: ios/PuzzleForge/SocialSharingComponents.swift
//
//  ShareableContent.swift
//  PuzzleForge
//
//  Created by Sushanth Tiruvaipati on 6/18/25.
//


import SwiftUI
import MessageUI

// MARK: - Data Classes
struct ShareableContent {
    let type: ShareType
    let title: String
    let message: String
    let hashtags: [String]
    let customData: [String: Any]
    
    init(type: ShareType, title: String, message: String, hashtags: [String] = [], customData: [String: Any] = [:]) {
        self.type = type
        self.title = title
        self.message = message
        self.hashtags = hashtags
        self.customData = customData
    }
}

enum ShareType {
    case puzzleResult
    case achievement
    case levelUp
    case streak
    case customPuzzle
    case weeklyRank
    case challengeCompletion
}

struct ShareOption {
    let title: String
    let iconName: String
    let color: Color
    let action: (ShareableContent) -> Void
}

// MARK: - Main Sharing Component
struct SocialShareDialog: View {
    let shareableContent: ShareableContent
    let showCustomPuzzleSharing: Bool
    let onDismiss: () -> Void
    
    @State private var showCustomDialog = false
    @State private var showShareSheet = false
    @State private var shareItems: [Any] = []
    
    var body: some View {
        VStack(spacing: 20) {
            // Title
            Text("Share Your Success!")
                .font(.title2)
                .fontWeight(.bold)
                .multilineTextAlignment(.center)
            
            // Preview of what will be shared
            VStack(spacing: 16) {
                RoundedRectangle(cornerRadius: 12)
                    .fill(Color(.systemGray6))
                    .frame(height: 120)
                    .overlay(
                        VStack(spacing: 8) {
                            Text(shareableContent.title)
                                .font(.headline)
                                .fontWeight(.bold)
                                .multilineTextAlignment(.center)
                            
                            Text(shareableContent.message)
                                .font(.subheadline)
                                .foregroundColor(.secondary)
                                .multilineTextAlignment(.center)
                        }
                        .padding(16)
                    )
                
                // Sharing options
                VStack(spacing: 8) {
                    ForEach(getShareOptions(), id: \.title) { option in
                        ShareOptionRow(
                            option: option,
                            content: shareableContent,
                            onClick: {
                                option.action(shareableContent)
                                onDismiss()
                            }
                        )
                    }
                    
                    // Custom puzzle sharing
                    if showCustomPuzzleSharing {
                        ShareOptionRow(
                            option: ShareOption(
                                title: "Send via Email",
                                iconName: "envelope.fill",
                                color: Color.blue,
                                action: { _ in }
                            ),
                            content: shareableContent,
                            onClick: {
                                showCustomDialog = true
                            }
                        )
                    }
                }
            }
            
            // Cancel button
            Button("Cancel") {
                onDismiss()
            }
            .foregroundColor(.secondary)
        }
        .padding(24)
        .background(Color(.systemBackground))
        .cornerRadius(20)
        .sheet(isPresented: $showShareSheet) {
            ShareSheet(activityItems: shareItems)
        }
        .sheet(isPresented: $showCustomDialog) {
            CustomPuzzleShareDialog(
                onDismiss: { showCustomDialog = false },
                onShare: { email, name in
                    let puzzleId = shareableContent.customData["puzzleId"] as? String ?? ""
                    sharePuzzle(puzzleId: puzzleId, recipientEmail: email, recipientName: name)
                    showCustomDialog = false
                    onDismiss()
                }
            )
        }
    }
    
    private func getShareOptions() -> [ShareOption] {
        return [
            ShareOption(
                title: "Share to Social Media",
                iconName: "square.and.arrow.up",
                color: Color.blue,
                action: { content in shareToSocialMedia(content) }
            ),
            ShareOption(
                title: "Copy to Clipboard",
                iconName: "doc.on.clipboard",
                color: Color.green,
                action: { content in copyToClipboard(content) }
            ),
            ShareOption(
                title: "Send via Messages",
                iconName: "message.fill",
                color: Color.blue,
                action: { content in shareViaMessages(content) }
            )
        ]
    }
    
    private func shareToSocialMedia(_ content: ShareableContent) {
        let shareText = buildShareText(content)
        shareItems = [shareText]
        showShareSheet = true
    }
    
    private func copyToClipboard(_ content: ShareableContent) {
        let shareText = buildShareText(content)
        UIPasteboard.general.string = shareText
        
        // Show toast-like feedback
        let impactFeedback = UIImpactFeedbackGenerator(style: .medium)
        impactFeedback.impactOccurred()
    }
    
    private func shareViaMessages(_ content: ShareableContent) {
        let shareText = buildShareText(content)
        
        if MFMessageComposeViewController.canSendText() {
            // Open Messages app directly
            let urlString = "sms:&body=\(shareText.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")"
            if let url = URL(string: urlString) {
                UIApplication.shared.open(url)
            }
        } else {
            // Fallback to general sharing
            shareToSocialMedia(content)
        }
    }
    
    private func sharePuzzle(puzzleId: String, recipientEmail: String, recipientName: String) {
        // Implementation for sharing custom puzzle via email
        print("Sharing puzzle \(puzzleId) to \(recipientEmail)")
    }
}

struct ShareOptionRow: View {
    let option: ShareOption
    let content: ShareableContent
    let onClick: () -> Void
    
    var body: some View {
        Button(action: onClick) {
            HStack(spacing: 12) {
                RoundedRectangle(cornerRadius: 20)
                    .fill(option.color.opacity(0.1))
                    .frame(width: 40, height: 40)
                    .overlay(
                        Image(systemName: option.iconName)
                            .foregroundColor(option.color)
                            .font(.system(size: 20))
                    )
                
                Text(option.title)
                    .font(.body)
                    .fontWeight(.medium)
                    .foregroundColor(.primary)
                
                Spacer()
                
                Image(systemName: "chevron.right")
                    .foregroundColor(.secondary)
                    .font(.system(size: 20))
            }
            .padding(.vertical, 8)
            .padding(.horizontal, 4)
        }
        .buttonStyle(PlainButtonStyle())
    }
}

struct CustomPuzzleShareDialog: View {
    let onDismiss: () -> Void
    let onShare: (String, String) -> Void
    
    @State private var recipientEmail = ""
    @State private var recipientName = ""
    
    var body: some View {
        NavigationView {
            VStack(spacing: 20) {
                Text("Share Puzzle")
                    .font(.title2)
                    .fontWeight(.bold)
                
                VStack(spacing: 16) {
                    VStack(alignment: .leading, spacing: 8) {
                        Text("Recipient Email")
                            .font(.subheadline)
                            .foregroundColor(.secondary)
                        TextField("Enter email address", text: $recipientEmail)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .keyboardType(.emailAddress)
                            .autocapitalization(.none)
                    }
                    
                    VStack(alignment: .leading, spacing: 8) {
                        Text("Recipient Name")
                            .font(.subheadline)
                            .foregroundColor(.secondary)
                        TextField("Enter name", text: $recipientName)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                    }
                }
                
                Spacer()
                
                HStack(spacing: 16) {
                    Button("Cancel") {
                        onDismiss()
                    }
                    .foregroundColor(.secondary)
                    .frame(maxWidth: .infinity)
                    
                    Button("Share") {
                        onShare(recipientEmail, recipientName)
                    }
                    .disabled(recipientEmail.isEmpty || recipientName.isEmpty)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 12)
                    .background(recipientEmail.isEmpty || recipientName.isEmpty ? Color.gray : Color.blue)
                    .foregroundColor(.white)
                    .cornerRadius(8)
                }
            }
            .padding(24)
            .navigationBarHidden(true)
        }
    }
}

// MARK: - Share Sheet for iOS
struct ShareSheet: UIViewControllerRepresentable {
    let activityItems: [Any]
    
    func makeUIViewController(context: Context) -> UIActivityViewController {
        let controller = UIActivityViewController(activityItems: activityItems, applicationActivities: nil)
        return controller
    }
    
    func updateUIViewController(_ uiViewController: UIActivityViewController, context: Context) {}
}

// MARK: - Content Builders
func buildShareText(_ content: ShareableContent) -> String {
    let hashtags = content.hashtags.isEmpty ? 
        " #PuzzleVerse #BrainTraining" : 
        " " + content.hashtags.map { "#\($0)" }.joined(separator: " ")
    
    switch content.type {
    case .puzzleResult:
        return "\(content.message)\(hashtags)"
    case .achievement:
        return "🏆 \(content.title)\n\(content.message)\(hashtags)"
    case .levelUp:
        return "📈 \(content.title)\n\(content.message)\(hashtags)"
    case .streak:
        return "🔥 \(content.title)\n\(content.message)\(hashtags)"
    case .weeklyRank:
        return "📊 \(content.title)\n\(content.message)\(hashtags)"
    case .challengeCompletion:
        return "✅ \(content.title)\n\(content.message)\(hashtags)"
    case .customPuzzle:
        return "🧩 \(content.title)\n\(content.message)\(hashtags)"
    }
}

// MARK: - Helper Functions for Creating Shareable Content
func createPuzzleResultShare(puzzleType: String, score: Int, streak: Int) -> ShareableContent {
    return ShareableContent(
        type: .puzzleResult,
        title: "Puzzle Completed!",
        message: "🧠 Just solved a \(puzzleType) puzzle and scored \(score) points! 🔥 Current streak: \(streak) days. Can you beat my score?",
        hashtags: ["PuzzleVerse", "BrainTraining", puzzleType.capitalized]
    )
}

func createAchievementShare(_ achievement: Achievement) -> ShareableContent {
    return ShareableContent(
        type: .achievement,
        title: "Achievement Unlocked!",
        message: "Just unlocked '\(achievement.title)' badge! \(achievement.description) 🎯",
        hashtags: ["PuzzleVerse", "Achievement", "BrainTraining"]
    )
}

func createLevelUpShare(oldLevel: Int, newLevel: Int) -> ShareableContent {
    return ShareableContent(
        type: .levelUp,
        title: "Level Up!",
        message: "🚀 Just reached Level \(newLevel) in PuzzleVerse! My brain training is paying off! 💪",
        hashtags: ["PuzzleVerse", "LevelUp", "BrainTraining"]
    )
}

func createStreakShare(streakDays: Int) -> ShareableContent {
    return ShareableContent(
        type: .streak,
        title: "Streak Achievement!",
        message: "🔥 \(streakDays) days of consistent brain training! Join me in daily puzzle solving! 🧠",
        hashtags: ["PuzzleVerse", "Streak", "DailyChallenge"]
    )
}

func createWeeklyRankShare(rank: Int, score: Int) -> ShareableContent {
    return ShareableContent(
        type: .weeklyRank,
        title: "Weekly Leaderboard!",
        message: "📊 This week I'm ranked #\(rank) on PuzzleVerse with \(score) points! 🧠 Join me in daily brain training!",
        hashtags: ["PuzzleVerse", "Leaderboard", "Competition"]
    )
}

func createCustomPuzzleShare(puzzleId: String, puzzleName: String) -> ShareableContent {
    return ShareableContent(
        type: .customPuzzle,
        title: "Custom Puzzle Created!",
        message: "🧩 I just created a custom puzzle '\(puzzleName)' on PuzzleVerse! Think you can solve it?",
        hashtags: ["PuzzleVerse", "CustomPuzzle", "Challenge"],
        customData: ["puzzleId": puzzleId]
    )
}

File: android/app/src/main/java/com/kreativekoala/riddleverse/SocialSharingComponents.kt
package com.kreativekoala.riddleverse

import com.kreativekoala.riddleverse.ui.theme.*
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.google.firebase.auth.FirebaseAuth

// MARK: - Data Classes
data class ShareableContent(
    val type: ShareType,
    val title: String,
    val message: String,
    val hashtags: List<String> = emptyList(),
    val customData: Map<String, Any> = emptyMap()
)

enum class ShareType {
    PUZZLE_RESULT,
    ACHIEVEMENT,
    LEVEL_UP,
    STREAK,
    CUSTOM_PUZZLE,
    WEEKLY_RANK,
    CHALLENGE_COMPLETION
}

data class ShareOption(
    val title: String,
    val icon: ImageVector,
    val color: Color,
    val action: (Context, ShareableContent) -> Unit
)

// MARK: - Main Sharing Component
@Composable
fun SocialShareDialog(
    shareableContent: ShareableContent,
    onDismiss: () -> Unit,
    showCustomPuzzleSharing: Boolean = false // For your existing custom puzzle sharing
) {
    val context = LocalContext.current
    var showCustomDialog by remember { mutableStateOf(false) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = {
            Text(
                "Share Your Success!",
                fontWeight = FontWeight.Bold,
                textAlign = TextAlign.Center
            )
        },
        text = {
            Column {
                // Preview of what will be shared
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(
                        containerColor = RvSurface
                    ),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(
                        modifier = Modifier.padding(16.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Text(
                            shareableContent.title,
                            fontWeight = FontWeight.Bold,
                            fontSize = 16.sp,
                            textAlign = TextAlign.Center
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(
                            shareableContent.message,
                            fontSize = 14.sp,
                            color = RvInkSoft,
                            textAlign = TextAlign.Center
                        )
                    }
                }

                Spacer(modifier = Modifier.height(16.dp))

                // Sharing options
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    getShareOptions().forEach { option ->
                        ShareOptionRow(
                            option = option,
                            content = shareableContent,
                            onClick = {
                                option.action(context, shareableContent)
                                onDismiss()
                            }
                        )
                    }

                    // Custom puzzle sharing (your existing feature)
                    if (showCustomPuzzleSharing) {
                        ShareOptionRow(
                            option = ShareOption(
                                title = stringResource(R.string.send_via_email),
                                icon = Icons.Default.Email,
                                color = RvSky,
                                action = { _, _ -> }
                            ),
                            content = shareableContent,
                            onClick = {
                                showCustomDialog = true
                            }
                        )
                    }
                }
            }
        },
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text(stringResource(R.string.cancel))
            }
        }
    )

    // Your existing custom puzzle dialog
    if (showCustomDialog) {
        CustomPuzzleShareDialog(
            onDismiss = { showCustomDialog = false },
            onShare = { email, name ->
                // Your existing sharePuzzle function
                val puzzleId = shareableContent.customData["puzzleId"] as? String ?: ""
                sharePuzzle(
                    puzzleId = puzzleId,
                    recipientEmail = email,
                    recipientName = name,
                    senderId = FirebaseAuth.getInstance().currentUser?.displayName
                        ?: FirebaseAuth.getInstance().currentUser?.email
                        ?: "anonymous"
                )
                showCustomDialog = false
                onDismiss()
            }
        )
    }
}

@Composable
fun ShareOptionRow(
    option: ShareOption,
    content: ShareableContent,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onClick() }
            .padding(vertical = 8.dp, horizontal = 4.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(
            modifier = Modifier
                .size(40.dp)
                .background(option.color.copy(alpha = 0.1f), CircleShape),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = option.icon,
                contentDescription = option.title,
                tint = option.color,
                modifier = Modifier.size(20.dp)
            )
        }

        Spacer(modifier = Modifier.width(12.dp))

        Text(
            text = option.title,
            fontSize = 16.sp,
            fontWeight = FontWeight.Medium,
            modifier = Modifier.weight(1f)
        )

        Icon(
            imageVector = Icons.Default.ChevronRight,
            contentDescription = "Share",
            tint = RvInkSoft,
            modifier = Modifier.size(20.dp)
        )
    }
}

@Composable
fun CustomPuzzleShareDialog(
    onDismiss: () -> Unit,
    onShare: (String, String) -> Unit
) {
    var recipientEmail by remember { mutableStateOf("") }
    var recipientName by remember { mutableStateOf("") }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(stringResource(R.string.share_puzzle)) },
        text = {
            Column {
                OutlinedTextField(
                    value = recipientEmail,
                    onValueChange = { recipientEmail = it },
                    label = { Text(stringResource(R.string.recipient_email)) },
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = recipientName,
                    onValueChange = { recipientName = it },
                    label = { Text(stringResource(R.string.recipient_name)) },
                    modifier = Modifier.fillMaxWidth()
                )
            }
        },
        confirmButton = {
            Button(
                onClick = { onShare(recipientEmail, recipientName) },
                enabled = recipientEmail.isNotBlank() && recipientName.isNotBlank()
            ) {
                Text(stringResource(R.string.share))
            }
        },
        dismissButton = {
            OutlinedButton(onClick = onDismiss) {
                Text(stringResource(R.string.cancel))
            }
        }
    )
}

// MARK: - Share Options Configuration
fun getShareOptions(): List<ShareOption> {
    return listOf(
        ShareOption(
            title = "Share to Social Media",
            icon = Icons.Default.Share,
            color = Color(0xFF1DA1F2),
            action = { context, content -> shareToSocialMedia(context, content) }
        ),
        ShareOption(
            title = "Copy to Clipboard",
            icon = Icons.Default.ContentCopy,
            color = Color(0xFF34A853),
            action = { context, content -> copyToClipboard(context, content) }
        ),
        ShareOption(
            title = "Send via Messages",
            icon = Icons.Default.Message,
            color = Color(0xFF0F9D58),
            action = { context, content -> shareViaMessages(context, content) }
        )
    )
}

// MARK: - Sharing Actions
fun shareToSocialMedia(context: Context, content: ShareableContent) {
    val shareText = buildShareText(content)
    val shareIntent = Intent().apply {
        action = Intent.ACTION_SEND
        type = "text/plain"
        putExtra(Intent.EXTRA_TEXT, shareText)
        putExtra(Intent.EXTRA_SUBJECT, content.title)
    }

    context.startActivity(Intent.createChooser(shareIntent, "Share your achievement"))
}

fun copyToClipboard(context: Context, content: ShareableContent) {
    val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as android.content.ClipboardManager
    val clip = android.content.ClipData.newPlainText("PuzzleVerse Achievement", buildShareText(content))
    clipboard.setPrimaryClip(clip)

    // Show toast (you might want to use a more elegant notification)
    android.widget.Toast.makeText(context, context.getString(R.string.copied_to_clipboard), android.widget.Toast.LENGTH_SHORT).show()
}

fun shareViaMessages(context: Context, content: ShareableContent) {
    val shareText = buildShareText(content)
    val intent = Intent(Intent.ACTION_SEND).apply {
        type = "text/plain"
        putExtra(Intent.EXTRA_TEXT, shareText)
        setPackage("com.android.mms") // Try to open default messaging app
    }

    try {
        context.startActivity(intent)
    } catch (e: Exception) {
        // Fallback to general sharing
        shareToSocialMedia(context, content)
    }
}

// MARK: - Content Builders
fun buildShareText(content: ShareableContent): String {
    val hashtags = if (content.hashtags.isNotEmpty()) {
        " " + content.hashtags.joinToString(" ") { "#$it" }
    } else {
        " #PuzzleVerse #BrainTraining"
    }

    return when (content.type) {
        ShareType.PUZZLE_RESULT -> {
            "${content.message}$hashtags"
        }
        ShareType.ACHIEVEMENT -> {
            "🏆 ${content.title}\n${content.message}$hashtags"
        }
        ShareType.LEVEL_UP -> {
            "📈 ${content.title}\n${content.message}$hashtags"
        }
        ShareType.STREAK -> {
            "🔥 ${content.title}\n${content.message}$hashtags"
        }
        ShareType.WEEKLY_RANK -> {
            "📊 ${content.title}\n${content.message}$hashtags"
        }
        ShareType.CHALLENGE_COMPLETION -> {
            "✅ ${content.title}\n${content.message}$hashtags"
        }
        ShareType.CUSTOM_PUZZLE -> {
            "🧩 ${content.title}\n${content.message}$hashtags"
        }
    }
}

// MARK: - Helper Functions for Creating Shareable Content
fun createPuzzleResultShare(puzzleType: String, score: Int, streak: Int): ShareableContent {
    return ShareableContent(
        type = ShareType.PUZZLE_RESULT,
        title = "Puzzle Completed!",
        message = "🧠 Just solved a $puzzleType puzzle and scored $score points! 🔥 Current streak: $streak days. Can you beat my score?",
        hashtags = listOf("PuzzleVerse", "BrainTraining", puzzleType.capitalize())
    )
}

fun createAchievementShare(achievement: Achievement): ShareableContent {
    return ShareableContent(
        type = ShareType.ACHIEVEMENT,
        title = "Achievement Unlocked!",
        message = "Just unlocked '${achievement.title}' badge! ${achievement.description} 🎯",
        hashtags = listOf("PuzzleVerse", "Achievement", "BrainTraining")
    )
}

fun createLevelUpShare(oldLevel: Int, newLevel: Int): ShareableContent {
    return ShareableContent(
        type = ShareType.LEVEL_UP,
        title = "Level Up!",
        message = "🚀 Just reached Level $newLevel in PuzzleVerse! My brain training is paying off! 💪",
        hashtags = listOf("PuzzleVerse", "LevelUp", "BrainTraining")
    )
}

fun createStreakShare(streakDays: Int): ShareableContent {
    return ShareableContent(
        type = ShareType.STREAK,
        title = "Streak Achievement!",
        message = "🔥 $streakDays days of consistent brain training! Join me in daily puzzle solving! 🧠",
        hashtags = listOf("PuzzleVerse", "Streak", "DailyChallenge")
    )
}

fun createWeeklyRankShare(rank: Int, score: Int): ShareableContent {
    return ShareableContent(
        type = ShareType.WEEKLY_RANK,
        title = "Weekly Leaderboard!",
        message = "📊 This week I'm ranked #$rank on PuzzleVerse with $score points! 🧠 Join me in daily brain training!",
        hashtags = listOf("PuzzleVerse", "Leaderboard", "Competition")
    )
}

fun createCustomPuzzleShare(puzzleId: String, puzzleName: String): ShareableContent {
    return ShareableContent(
        type = ShareType.CUSTOM_PUZZLE,
        title = "Custom Puzzle Created!",
        message = "🧩 I just created a custom puzzle '$puzzleName' on PuzzleVerse! Think you can solve it?",
        hashtags = listOf("PuzzleVerse", "CustomPuzzle", "Challenge"),
        customData = mapOf("puzzleId" to puzzleId)
    )
}